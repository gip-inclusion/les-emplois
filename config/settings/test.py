import os


ITOU_ENVIRONMENT = "DEV"
os.environ["ITOU_ENVIRONMENT"] = ITOU_ENVIRONMENT

from .base import *  # noqa: E402,F403


SECRET_KEY = "foobar"
ALLOWED_HOSTS = []

# We *want* to do the same `collectstatic` on the CI than on PROD to catch errors early,
# but we don't want to do it when running the test suite locally for performance reasons.
if not os.getenv("CI", False):
    # `ManifestStaticFilesStorage` (used in base settings) requires `collectstatic` to be run.
    STORAGES["staticfiles"]["BACKEND"] = "django.contrib.staticfiles.storage.StaticFilesStorage"  # noqa: F405

ASP_ITOU_PREFIX = "XXXXX"  # same as in our fixtures
ITOU_PROTOCOL = "http"
ITOU_FQDN = "localhost:8000"

DATABASES["default"]["HOST"] = os.getenv("PGHOST", "127.0.0.1")  # noqa: F405
DATABASES["default"]["PORT"] = os.getenv("PGPORT", "5432")  # noqa: F405
DATABASES["default"]["NAME"] = os.getenv("PGDATABASE", "itou")  # noqa: F405
DATABASES["default"]["USER"] = os.getenv("PGUSER", "postgres")  # noqa: F405
DATABASES["default"]["PASSWORD"] = os.getenv("PGPASSWORD", "password")  # noqa: F405

MAILJET_API_KEY_PRINCIPAL = "API_MAILJET_KEY_PRINCIPAL"
MAILJET_SECRET_KEY_PRINCIPAL = "API_MAILJET_SECRET_PRINCIPAL"

AWS_S3_ENDPOINT_URL = f"http://{os.getenv('CELLAR_ADDON_HOST', 'localhost:9000')}/"
AWS_S3_ACCESS_KEY_ID = "minioadmin"
AWS_S3_SECRET_ACCESS_KEY = "minioadmin"
AWS_STORAGE_BUCKET_NAME = "tests"
