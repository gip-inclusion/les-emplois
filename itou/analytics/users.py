from django.db.models import Count

from itou.users import enums as users_enums, models as users_models

from . import models


def collect_analytics_data(before):
    counts = {
        info["kind"]: info["count"]
        for info in users_models.User.objects.values("kind").annotate(count=Count("pk")).values("kind", "count")
    }
    return {
        models.DatumCode.USER_COUNT: sum(counts.values()),
        models.DatumCode.USER_JOB_SEEKER_COUNT: counts.get(users_enums.KIND_JOB_SEEKER, 0),
        models.DatumCode.USER_PRESCRIBER_COUNT: counts.get(users_enums.KIND_PRESCRIBER, 0),
        models.DatumCode.USER_EMPLOYER_COUNT: counts.get(users_enums.KIND_EMPLOYER, 0),
        models.DatumCode.USER_LABOR_INSPECTOR_COUNT: counts.get(users_enums.KIND_LABOR_INSPECTOR, 0),
        models.DatumCode.USER_ITOU_STAFF_COUNT: counts.get(users_enums.KIND_ITOU_STAFF, 0),
    }
