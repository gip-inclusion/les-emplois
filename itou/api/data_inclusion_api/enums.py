import enum


class StructureTypeStr(str, enum.Enum):
    ORGA = "orga"
    SIAE = "siae"
