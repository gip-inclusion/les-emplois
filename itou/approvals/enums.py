from django.db import models

from itou.companies import enums as companies_enums


class ApprovalStatus(models.TextChoices):
    EXPIRED = "EXPIRED", "Expiré"
    VALID = "VALID", "Valide"
    FUTURE = "FUTURE", "Valide (non démarré)"
    SUSPENDED = "SUSPENDED", "Valide (suspendu)"


class Origin(models.TextChoices):
    DEFAULT = "default", "Créé normalement via les emplois"
    PE_APPROVAL = "pe_approval", "Créé lors d'un import d'Agrément Pole Emploi"
    # On November 30th, 2021, AI were delivered approvals without a diagnosis.
    AI_STOCK = "ai_stock", "Créé lors de l'import du stock AI"
    ADMIN = "admin", "Créé depuis l'admin"


class ProlongationReason(models.TextChoices):
    SENIOR_CDI = "SENIOR_CDI", "CDI conclu avec une personne de plus de 57 ans"
    COMPLETE_TRAINING = "COMPLETE_TRAINING", "Fin d'une formation"
    RQTH = "RQTH", "RQTH - Reconnaissance de la qualité de travailleur handicapé"
    SENIOR = "SENIOR", "50 ans et plus"
    # Exclusive to AI and ACI
    PARTICULAR_DIFFICULTIES = (
        "PARTICULAR_DIFFICULTIES",
        "Difficultés particulièrement importantes dont l'absence de prise en charge ferait "
        "obstacle à son insertion professionnelle",
    )
    # Since December 1, 2021, health context reason can no longer be used
    HEALTH_CONTEXT = "HEALTH_CONTEXT", "Contexte sanitaire"

    @classmethod
    def for_company(cls, company):
        enums = [
            cls.SENIOR_CDI,
            cls.COMPLETE_TRAINING,
            cls.RQTH,
            cls.SENIOR,
        ]
        if company.kind in [companies_enums.CompanyKind.AI, companies_enums.CompanyKind.ACI]:
            enums.append(cls.PARTICULAR_DIFFICULTIES)

        empty = [(None, cls.__empty__)] if hasattr(cls, "__empty__") else []
        return empty + [(enum.value, enum.label) for enum in enums]


class ProlongationRequestStatus(models.TextChoices):
    PENDING = "PENDING", "À traiter"
    GRANTED = "GRANTED", "Acceptée"
    DENIED = "DENIED", "Refusée"


class ProlongationRequestDenyReason(models.TextChoices):
    IAE = "IAE", "L’IAE ne correspond plus aux besoins / à la situation de la personne."
    SIAE = "SIAE", "La typologie de SIAE ne correspond plus aux besoins / à la situation de la personne."
    DURATION = "DURATION", "La durée de prolongation demandée n’est pas adaptée à la situation du candidat."


class ProlongationRequestDenyProposedAction(models.TextChoices):
    # TODO: Clarify the actions to improve the naming
    EXIT_IAE = (
        "EXIT_IAE",
        "Accompagnement à la recherche d’emploi hors IAE et mobilisation de l’offre "
        "de services disponible au sein de votre structure ou celle d’un partenaire.",
    )
    SOCIAL_PARTNER = "SOCIAL_PARTNER", "Orientation vers un partenaire de l’accompagnement social/professionnel."
    OTHER = "OTHER", "Autre"
