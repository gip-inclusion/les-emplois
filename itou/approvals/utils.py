from itou.companies.enums import SIAE_WITH_CONVENTION_KINDS


def get_user_last_accepted_siae_job_application(user):
    if not user.is_job_seeker:
        return None

    # Some candidates may not have accepted job applications
    # Assuming it's the case can lead to issues downstream
    return (
        user.job_applications.accepted()
        .filter(to_company__kind__in=SIAE_WITH_CONVENTION_KINDS)
        .with_accepted_at()
        .order_by("-accepted_at", "-hiring_start_at")
        .first()
    )


def last_hire_was_made_by_siae(user, siae):
    if not user.is_job_seeker:
        return False
    last_accepted_job_application = get_user_last_accepted_siae_job_application(user)
    return last_accepted_job_application and last_accepted_job_application.to_company_id == siae.pk
