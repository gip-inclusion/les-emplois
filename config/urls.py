from django.conf import settings
from django.contrib import admin
from django.urls import include, path, re_path, register_converter
from django.views.generic import TemplateView

from itou.utils.urls import SiretConverter
from itou.www.dashboard import views as dashboard_views
from itou.www.login import views as login_views
from itou.www.signup import views as signup_views


register_converter(SiretConverter, "siret")

urlpatterns = [
    path("admin/", admin.site.urls),
    # --------------------------------------------------------------------------------------
    # allauth URLs. Order is important because some URLs are overriden.
    # --------------------------------------------------------------------------------------
    # Override allauth `account_signup` URL.
    # /accounts/signup/ <=> account_signup
    # We don't want any user to be able to signup using the default allauth `signup` url
    # because we have multiple specific signup processes for different kind of users.
    re_path(r"^accounts/signup/$", signup_views.signup),
    # --------------------------------------------------------------------------------------
    # Override allauth `account_login` URL.
    # /accounts/login/ <=> account_login
    # We override this view because the login page should look slightly differently
    # for job seekers, prescribers and employers.
    # Also, PEAMU is only available for job seekers.
    re_path(r"^accounts/login/$", login_views.login),
    # --------------------------------------------------------------------------------------
    # Override allauth `account_change_password` URL.
    # /accounts/password/change/ <=> account_change_password
    # https://github.com/pennersr/django-allauth/issues/468
    re_path(r"^accounts/password/change/$", dashboard_views.password_change),
    # --------------------------------------------------------------------------------------
    # Override allauth `account_reset_password` URL.
    # Avoid user enumeration via password reset page.
    re_path(r"^accounts/password/reset/$", signup_views.ItouPasswordResetView.as_view()),
    # --------------------------------------------------------------------------------------
    # Override allauth `account_logout` URL.
    # /accounts/logout/ <=> account_logout
    # We need custom code to process PEAMU logout.
    re_path(r"^accounts/logout/$", dashboard_views.logout),
    # --------------------------------------------------------------------------------------
    # Other allauth URLs.
    path("accounts/", include("allauth.urls")),
    # --------------------------------------------------------------------------------------
    # PEAMU URLs.
    path("accounts/", include("itou.allauth_adapters.peamu.urls")),
    # FranceConnect URLs.
    path("franceconnect/", include("itou.france_connect.urls")),
    # --------------------------------------------------------------------------------------
    # API.
    path("api/v1/", include("itou.api.urls", namespace="v1")),
    path("api/", include("itou.www.api.urls")),
    # www.
    path("", include("itou.www.home.urls")),
    path("apply/", include("itou.www.apply.urls")),
    path("approvals/", include("itou.www.approvals_views.urls")),
    path("autocomplete/", include("itou.www.autocomplete.urls")),
    path("dashboard/", include("itou.www.dashboard.urls")),
    path("employee_record/", include("itou.www.employee_record_views.urls")),
    path("institutions/", include("itou.www.institutions_views.urls")),
    path("invitations/", include("itou.www.invitations_views.urls")),
    path("prescribers/", include("itou.www.prescribers_views.urls")),
    path("search/", include("itou.www.search.urls")),
    path("siae/", include("itou.www.siaes_views.urls")),
    path("signup/", include("itou.www.signup.urls")),
    path("stats/", include("itou.www.stats.urls")),
    path("versions/", include("itou.www.releases.urls")),
    # Enable Anymailâ€™s status tracking
    # https://anymail.readthedocs.io/en/stable/esps/mailjet/#status-tracking-webhooks
    re_path(r"^webhooks/anymail/", include("anymail.urls")),
    path("welcoming_tour/", include("itou.www.welcoming_tour.urls")),
    # Static pages.
    path("accessibility/", TemplateView.as_view(template_name="static/accessibility.html"), name="accessibility"),
    # Errors pages.
    path("403/", TemplateView.as_view(template_name="403.html"), name="403"),
    path("404/", TemplateView.as_view(template_name="404.html"), name="404"),
    path("500/", TemplateView.as_view(template_name="500.html"), name="500"),
]

if settings.DEBUG and "debug_toolbar" in settings.INSTALLED_APPS:
    import debug_toolbar

    urlpatterns = [path("__debug__/", include(debug_toolbar.urls))] + urlpatterns
    # In the local developpemnt environment, we enable the default callback URLs
    urlpatterns = [path("", include("itou.france_connect.urls"))] + urlpatterns
