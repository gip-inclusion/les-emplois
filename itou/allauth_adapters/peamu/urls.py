from django.urls import include, re_path

from itou.allauth_adapters.peamu import views as peamu_views
from itou.allauth_adapters.peamu.provider import PEAMUProvider


def default_urlpatterns(provider):
    urlpatterns = [
        re_path(r"^login/$", peamu_views.oauth2_login, name=provider.id + "_login"),
        re_path(r"^login/callback/$", peamu_views.oauth2_callback, name=provider.id + "_callback"),
    ]

    return [
        re_path(r"^profile/$", peamu_views.redirect_to_dashboard_view),
        re_path(r"^" + provider.get_slug() + r"/", include(urlpatterns)),
    ]


urlpatterns = default_urlpatterns(PEAMUProvider)
