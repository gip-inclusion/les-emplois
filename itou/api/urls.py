from django.urls import include, path
from drf_spectacular.views import SpectacularAPIView, SpectacularRedocView
from rest_framework import routers
from rest_framework.authtoken import views as auth_views

from itou.api.data_inclusion_api.views import DataInclusionStructureView

from .applicants_api.views import ApplicantsView
from .employee_record_api.viewsets import (
    DummyEmployeeRecordViewSet,
    EmployeeRecordUpdateNotificationViewSet,
    EmployeeRecordViewSet,
)
from .siae_api.viewsets import SiaeViewSet


# High level app for API
# https://docs.djangoproject.com/en/dev/topics/http/urls/#url-namespaces-and-included-urlconfs
app_name = "itou.api"

# Using DRF router with viewsets means automatic definition of utl patterns
router = routers.DefaultRouter()
router.register(r"employee-records", EmployeeRecordViewSet, basename="employee-records")
router.register(
    r"employee-record-notifications",
    EmployeeRecordUpdateNotificationViewSet,
    basename="employee-record-notifications",
)
router.register(r"dummy-employee-records", DummyEmployeeRecordViewSet, basename="dummy-employee-records")
router.register(r"siaes", SiaeViewSet, basename="siaes")

urlpatterns = [
    # TokenAuthentication endpoint to get token from login/password.
    path("token-auth/", auth_views.obtain_auth_token, name="token-auth"),
    # Needed for Browseable API (dev)
    path("api-auth/", include("rest_framework.urls", namespace="rest_framework")),
    # OpenAPI
    # See: https://www.django-rest-framework.org/topics/documenting-your-api/
    # OAS 3 YAML schema (downloadable)
    path(
        "oas3/",
        SpectacularAPIView.as_view(),
        name="openapi_schema",
    ),
    path(
        "redoc/",
        SpectacularRedocView.as_view(url_name="v1:openapi_schema"),
        name="redoc",
    ),
]

urlpatterns += router.urls

urlpatterns += [
    path("structures/", DataInclusionStructureView.as_view(), name="structures-list"),
    path("candidats/", ApplicantsView.as_view(), name="applicants-list"),
]
