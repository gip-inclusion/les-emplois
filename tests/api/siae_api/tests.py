import json

from django.urls import reverse
from rest_framework.test import APIClient, APITestCase

from itou.companies.enums import CompanyKind
from tests.cities.factories import create_city_guerande, create_city_saint_andre
from tests.companies.factories import SiaeFactory
from tests.users.factories import EmployerFactory
from tests.utils.test import BASE_NUM_QUERIES


ENDPOINT_URL = reverse("v1:siaes-list")


class SiaeAPIFetchListTest(APITestCase):
    def setUp(self):
        super().setUp()
        self.client = APIClient()
        self.user = EmployerFactory()

        # We create 2 cities and 2 siaes in Saint-Andre.
        self.saint_andre = create_city_saint_andre()
        self.guerande = create_city_guerande()
        self.siae_a = SiaeFactory(kind=CompanyKind.EI, department="44", coords=self.saint_andre.coords)
        self.siae_b = SiaeFactory(
            with_jobs=True, romes=("N1101", "N1105", "N1103", "N4105"), department="44", coords=self.saint_andre.coords
        )

    def test_performances(self):
        num_queries = BASE_NUM_QUERIES
        num_queries += 1  # Get city with insee_code
        num_queries += 1  # Count siaes
        num_queries += 1  # Select sias
        num_queries += 1  # prefetch job_description_through
        with self.assertNumQueries(num_queries):
            query_params = {"code_insee": self.saint_andre.code_insee, "distance_max_km": 100}
            self.client.get(ENDPOINT_URL, query_params, format="json")

    def test_fetch_siae_list_without_params(self):
        """
        The query parameters for INSEE code and distance are mandatories
        """
        response = self.client.get(ENDPOINT_URL, format="json")

        self.assertContains(
            response, "Les paramètres `code_insee` et `distance_max_km` sont obligatoires.", status_code=400
        )

    def test_fetch_siae_list_with_too_high_distance(self):
        """
        The query parameter distance must be <= 100
        """
        query_params = {"code_insee": 44056, "distance_max_km": 200}
        response = self.client.get(ENDPOINT_URL, query_params, format="json")

        self.assertContains(
            response, "Le paramètre `distance_max_km` doit être compris entre 0 et 100", status_code=400
        )

    def test_fetch_siae_list_with_negative_distance(self):
        """
        The query parameter distance must be positive
        """
        query_params = {"code_insee": 44056, "distance_max_km": -10}
        response = self.client.get(ENDPOINT_URL, query_params, format="json")

        self.assertContains(
            response, "Le paramètre `distance_max_km` doit être compris entre 0 et 100", status_code=400
        )

    def test_fetch_siae_list_with_invalid_code_insee(self):
        """
        The query parameters for INSEE code and distance are mandatories
        """
        query_params = {"code_insee": 12345, "distance_max_km": 10}
        response = self.client.get(ENDPOINT_URL, query_params, format="json")

        assert response.content == b'{"detail":"Pas de ville avec pour code_insee 12345"}'
        assert response.status_code == 404

    def test_fetch_siae_list(self):
        """
        Search for siaes in the city that has 2 SIAES
        """

        query_params = {"code_insee": self.saint_andre.code_insee, "distance_max_km": 100}
        response = self.client.get(ENDPOINT_URL, query_params, format="json")

        body = json.loads(response.content)
        assert body["count"] == 2
        assert response.status_code == 200

    def test_fetch_siae_list_too_far(self):
        """
        Search for siaes in a city that has no SIAES
        """

        query_params = {"code_insee": self.guerande.code_insee, "distance_max_km": 10}
        response = self.client.get(ENDPOINT_URL, query_params, format="json")

        body = json.loads(response.content)
        assert body["count"] == 0
        assert response.status_code == 200
