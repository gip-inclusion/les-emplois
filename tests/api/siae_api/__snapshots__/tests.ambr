# serializer version: 1
# name: TestSiaeAPIFetchList.test_performances
  dict({
    'num_queries': 6,
    'queries': list([
      dict({
        'origin': list([
          'Atomic.__enter__[<site-packages>/django/db/transaction.py]',
        ]),
        'sql': 'SAVEPOINT "<snapshot>"',
      }),
      dict({
        'origin': list([
          'CompanyFilterSet.filter_queryset[api/siae_api/viewsets.py]',
          'SiaeViewSet.list[<site-packages>/rest_framework/mixins.py]',
          'SiaeViewSet.list[api/siae_api/viewsets.py]',
        ]),
        'sql': '''
          SELECT "cities_city"."id",
                 "cities_city"."name",
                 "cities_city"."slug",
                 "cities_city"."department",
                 "cities_city"."post_codes",
                 "cities_city"."code_insee",
                 "cities_city"."coords",
                 "cities_city"."edition_mode"
          FROM "cities_city"
          WHERE "cities_city"."code_insee" = %s
          LIMIT 21
        ''',
      }),
      dict({
        'origin': list([
          'Paginator.count[<site-packages>/django/core/paginator.py]',
          'SiaeViewSet.list[<site-packages>/rest_framework/mixins.py]',
          'SiaeViewSet.list[api/siae_api/viewsets.py]',
        ]),
        'sql': '''
          SELECT COUNT(*) AS "__count"
          FROM "companies_company"
          WHERE (NOT ("companies_company"."siret" = %s)
                 AND ST_DWithin("companies_company"."coords", %s, %s))
        ''',
      }),
      dict({
        'origin': list([
          'SiaeViewSet.list[<site-packages>/rest_framework/mixins.py]',
          'SiaeViewSet.list[api/siae_api/viewsets.py]',
        ]),
        'sql': '''
          SELECT "companies_company"."id",
                 "companies_company"."address_line_1",
                 "companies_company"."address_line_2",
                 "companies_company"."post_code",
                 "companies_company"."city",
                 "companies_company"."department",
                 "companies_company"."coords",
                 "companies_company"."geocoding_score",
                 "companies_company"."geocoding_updated_at",
                 "companies_company"."ban_api_resolved_address",
                 "companies_company"."insee_city_id",
                 "companies_company"."name",
                 "companies_company"."created_at",
                 "companies_company"."updated_at",
                 "companies_company"."uid",
                 "companies_company"."active_members_email_reminder_last_sent_at",
                 "companies_company"."siret",
                 "companies_company"."naf",
                 "companies_company"."kind",
                 "companies_company"."brand",
                 "companies_company"."phone",
                 "companies_company"."email",
                 "companies_company"."auth_email",
                 "companies_company"."website",
                 "companies_company"."description",
                 "companies_company"."provided_support",
                 "companies_company"."source",
                 "companies_company"."created_by_id",
                 "companies_company"."block_job_applications",
                 "companies_company"."job_applications_blocked_at",
                 "companies_company"."convention_id",
                 "companies_company"."job_app_score",
                 "companies_company"."is_searchable",
                 "companies_company"."rdv_solidarites_id"
          FROM "companies_company"
          WHERE (NOT ("companies_company"."siret" = %s)
                 AND ST_DWithin("companies_company"."coords", %s, %s))
          ORDER BY "companies_company"."id" ASC
          LIMIT 2
        ''',
      }),
      dict({
        'origin': list([
          'SiaeViewSet.list[<site-packages>/rest_framework/mixins.py]',
          'SiaeViewSet.list[api/siae_api/viewsets.py]',
        ]),
        'sql': '''
          SELECT "companies_jobdescription"."id",
                 "companies_jobdescription"."appellation_id",
                 "companies_jobdescription"."company_id",
                 "companies_jobdescription"."created_at",
                 "companies_jobdescription"."updated_at",
                 "companies_jobdescription"."is_active",
                 "companies_jobdescription"."custom_name",
                 "companies_jobdescription"."description",
                 "companies_jobdescription"."ui_rank",
                 "companies_jobdescription"."contract_type",
                 "companies_jobdescription"."other_contract_type",
                 "companies_jobdescription"."contract_nature",
                 "companies_jobdescription"."location_id",
                 "companies_jobdescription"."hours_per_week",
                 "companies_jobdescription"."open_positions",
                 "companies_jobdescription"."profile_description",
                 "companies_jobdescription"."is_resume_mandatory",
                 "companies_jobdescription"."is_qpv_mandatory",
                 "companies_jobdescription"."market_context_description",
                 "companies_jobdescription"."source_id",
                 "companies_jobdescription"."source_kind",
                 "companies_jobdescription"."source_url",
                 "companies_jobdescription"."field_history",
                 "companies_jobdescription"."creation_source",
                 "jobs_appellation"."updated_at",
                 "jobs_appellation"."code",
                 "jobs_appellation"."name",
                 "jobs_appellation"."rome_id",
                 "jobs_appellation"."full_text",
                 "jobs_rome"."updated_at",
                 "jobs_rome"."code",
                 "jobs_rome"."name",
                 "cities_city"."id",
                 "cities_city"."name",
                 "cities_city"."slug",
                 "cities_city"."department",
                 "cities_city"."post_codes",
                 "cities_city"."code_insee",
                 "cities_city"."coords",
                 "cities_city"."edition_mode"
          FROM "companies_jobdescription"
          INNER JOIN "jobs_appellation" ON ("companies_jobdescription"."appellation_id" = "jobs_appellation"."code")
          LEFT OUTER JOIN "jobs_rome" ON ("jobs_appellation"."rome_id" = "jobs_rome"."code")
          LEFT OUTER JOIN "cities_city" ON ("companies_jobdescription"."location_id" = "cities_city"."id")
          WHERE ("companies_jobdescription"."is_active"
                 AND "companies_jobdescription"."company_id" IN (%s,
                                                                 %s))
          ORDER BY "companies_jobdescription"."updated_at" DESC,
                   "companies_jobdescription"."created_at" DESC
        ''',
      }),
      dict({
        'origin': list([
          'Atomic.__exit__[<site-packages>/django/db/transaction.py]',
        ]),
        'sql': 'RELEASE SAVEPOINT "<snapshot>"',
      }),
    ]),
  })
# ---
